store/src/java/com/zimbra/cs/contacts/ContactAffinityQuery.java

```diff
@@ -0,0 +1,388 @@
+                String senderEmail = sender.getContactEmail();
+                List<AffinityTarget> others = targets.stream().filter(
+                        target -> !target.getContactEmail().equals(senderEmail)).collect(Collectors.toList());
+                TupleStream stream = buildSearchStream(others, EventType.AFFINITY, ignoreTargetAffinityField, senderEmail);
+                toMerge[idx] = stream;
+                idx++;
+            }
+            searchStream = mergeStreamsOnMessageId(toMerge);
+        }
+        return searchStream;
+    }
+
+    private TupleStream buildAffinityStream(AffinityScope scope) throws IOException {
+        EventType eventType;
+        TupleStream stream;
+
+        // 1. search for IDs of messages that satisfy the scope parameters
+        boolean ignoreTargetAffinityField = scope.getIgnoreTargetAffinityField();
+        switch (scope.getRelatedVia()) {
+        case OUTGOING_MSGS:
+            eventType = EventType.SENT;
+            boolean matchAllTargets = scope.getMatchType() == MatchType.ALL;
+            stream = getOutgoingMessageIds(matchAllTargets, ignoreTargetAffinityField);
+            break;
+        case INCOMING_MSGS:
+        default:
+            eventType = EventType.AFFINITY;
+            stream = getIncomingMessageIds(scope.getSenderMustBeTarget(), ignoreTargetAffinityField);
+            break;
+        }
+        // 2. walk to all other contacts referenced on these messages, counting the number of edges going to each contact
+        stream = addNodesStream(stream, eventType);
+
+        // 4. de-duplicate contacts
+        stream = addDedupeStream(stream);
+
+        // 5. sort by count
+        stream = addRankStream(stream);
+        return stream;
+    }
+
+    private void setStreamContext(TupleStream stream) {
+        StreamContext context = new StreamContext();
+        SolrClientCache clientCache = solrHelper.getClientCache();
+        context.setSolrClientCache(clientCache);
+        context.setModelCache(new ModelCache(MODEL_CACHE_SIZE, solrHelper.getZkHost(), clientCache));
+        stream.setStreamContext(context);
+
+    }
+
+    private void openStream(TupleStream stream) throws ServiceException {
+        try {
+            stream.open();
+        } catch (IOException e) {
+            throw ServiceException.FAILURE("error opening contact affinity TupleStream", e);
+        }
+    }
+
+    private Tuple readStream(TupleStream stream) throws ServiceException {
+        try {
+            return stream.read();
+        } catch (IOException e) {
+            throw ServiceException.FAILURE("error reading tuple from contact affinity TupleStream", e);
+        }
+    }
+
+    /**
+     * Compose and run a streaming expression to return related contacts for a given AffinityScope.
+     */
+    @VisibleForTesting
+    public RelatedContactsResults execute(AffinityScope scope) throws ServiceException {
+        try(TupleStream stream = buildAffinityStream(scope)) {
```

Comment:
Is the tupleStream being closed? I don't see that it is implementing the AutoClosable interface.

Reply:
`TupleStream` implements `Closeable`, which extends `AutoCloseable`.